c7n/filters/core.py

```diff
@@ -1018,3 +1020,142 @@ def reorder(self, items, key=None):
+
+    def __init__(self, *args, **kw):
+        super(FilterRegistry, self).__init__(*args, **kw)
+        self.register('value', ValueFilter)
+        self.register('or', Or)
+        self.register('and', And)
+        self.register('not', Not)
+        self.register('event', EventFilter)
+        self.register('reduce', ReduceFilter)
+
+
+class ListItemResourceManager(ResourceManager):
+    filter_registry = ListItemRegistry('filters')
+
+    def get_model(self):
+        return ListItemModel()
+
+
+class ListItemFilter(Filter):
+    """
+    Perform multi attribute filtering on items within a list,
+    for example looking for security groups that have rules which
+    include 0.0.0.0/0 and port 22 open.
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: security-group-with-22-open-to-world
+            resource: aws.security-group
+            filters:
+              - type: list-item
+                key: IpPermissions
+                attrs:
+                  - type: value
+                    key: IpRanges[].CidrIp
+                    value: '0.0.0.0/0'
+                    op: in
+                    value_type: swap
+                  - type: value
+                    key: FromPort
+                    value: 22
+                  - type: value
+                    key: ToPort
+                    value: 22
+          - name: find-task-def-not-using-registry
+            resource: aws.ecs-task-definition
+            filters:
+              - not:
+                - type: list-item
+                  key: containerDefinitions
+                  attrs:
+                    - not:
+                      - type: value
+                        key: image
+                        value: "${account_id}.dkr.ecr.us-east-2.amazonaws.com.*"
+                        op: regex
+    """
+
+    def _get_schema():
+        base_filters = [
+            {'$ref': '#/definitions/filters/event'},
+            {'$ref': '#/definitions/filters/reduce'},
+            {'$ref': '#/definitions/filters/value'},
+            {'$ref': '#/definitions/filters/valuekv'},
+            {
+                'enum': [
+                    'value',
+                    'or',
+                    'and',
+                    'not',
+                    'event',
+                    'reduce',
+                ]
+            }
+        ]
+        any_of = []
+        any_of.extend(base_filters)
+
+        for i in ('and', 'or', 'not',):
+            any_of.append(
+                {
+                    'additional_properties': False,
+                    'properties': {
+                        i: {
+                            'type': 'array',
+                            'items': {
+                                'anyOf': base_filters
+                            }
+                        }
+                    },
+                    'type': 'object'
+                }
+            )
+
+        schema = {
```

Comment:
I feel like we could use additional attributes here re count of items that match, say cardinality: 2 (or perhaps set-size?) which defaults to equality but could also specify operator.